File: kaprekar.py
def largest_number(n):
    return int(''.join(sorted(str(n), reverse=True)))

def smallest_number(n):
    return int(''.join(sorted(str(n))))

def kaprekar(a):
    largest = largest_number(a)
    smallest = smallest_number(a)
    return largest, smallest, largest - smallest

def all_digits_same(n):
    return len(set(str(n))) == 1

def kaprekar_p(p):
    return p == 6174

for i in range(1000, 10000):
    if all_digits_same(i):
        continue

    attempts = 1
    current_value = i
    while True:
        largest, smallest, k = kaprekar(current_value)
        print(f"{i}. Attempt #{attempts}.  {largest}-{smallest}={k}{'*' if kaprekar_p(k) else ''}")
        if kaprekar_p(k):
            break
        attempts += 1
        current_value = k  # Update current_value to k
        if all_digits_same(current_value):  # Check if all digits are the same
            break

File: kaprekar.rb

# This function takes an integer n, converts it to a 4-digit string with leading zeros if necessary,
# sorts the digits in descending order to form the largest possible number,
# and returns the resulting integer.

def largest_number(n)
    n.to_s.rjust(4, '0').split('').sort { |a, b| b + a <=> a + b }.join.to_i
end
# This function takes an integer n, converts it to a 4-digit string with leading zeros if necessary,
# sorts the digits in ascending order to form the smallest possible number,
# and returns the resulting integer.
  def smallest_number(n)
    n.to_s.rjust(4, '0').split('').sort.join.to_i
  end

  def kaprekar(a)
    largest, smallest = [largest_number(a), smallest_number(a)]
    [largest, smallest, (largest - smallest)]
  end

  def kaprekar?(p)
    p == 6174
  end

 
  def all_digits_same?(n)
   n.to_s.rjust(4, '0').chars.uniq.size == 1
  end

  (1000..9999).each do |i|
    next if all_digits_same?(i)
  
    attempts = 1
    current_value = i
    loop do
      largest, smallest, k = kaprekar(current_value)
      puts "#{i}. Attempt ##{attempts}.  #{largest}-#{smallest}=#{k}#{kaprekar?(k) ? "*\n" : ""}"
      if kaprekar?(k)
        puts "\n"
        break
      end
      attempts += 1
      current_value = k
    end
  end
